### PR TITLE
json loads 'stdout' and 'stderr' in shell & ssh runners

### PR DESCRIPTION
**Pull request by lakshmi-kannan**
* Add a json_loads utility method in st2common/util/jsonify.py
* Add test cases for utility method
* FabricRunner and LocalShellRunner to use utility method to transform
  stdout and stderr streams

## Sample output

```
(virtualenv)~/stanley git:STORM-931/parse_json_in_results_shell_runners ❯❯❯ st2 run core.local shit                                                                                                                             ✭ ✱ ◼
.
+-----------------+-----------------------------+
| Property        | Value                       |
+-----------------+-----------------------------+
| id              | 54aee0740640fd77c6c9f312    |
| context         | {                           |
|                 |     "user": "lakshmi"       |
|                 | }                           |
| parameters      | {                           |
|                 |     "cmd": "shit"           |
|                 | }                           |
| status          | succeeded                   |
| start_timestamp | 2015-01-08T19:54:28.122000Z |
| result          | {                           |
|                 |     "failed": false,        |
|                 |     "stderr": "",           |
|                 |     "return_code": 0,       |
|                 |     "succeeded": true,      |
|                 |     "stdout": {             |
|                 |         "foo": {            |
|                 |             "bar": "baz"    |
|                 |         }                   |
|                 |     }                       |
|                 | }                           |
| action          | core.local                  |
| callback        |                             |
| end_timestamp   | 2015-01-08T19:54:28.216000Z |
+-----------------+-----------------------------+
(virtualenv)~/stanley git:STORM-931/parse_json_in_results_shell_runners ❯❯❯
```

### CODE REVIEW

**Comment by jfryman** [comment]
Is there any way that it could not be `stdout` if the parsing happens? `result` like the python runner?

File: st2common/st2common/util/jsonify.py [hunk 1]
```diff
@@ -34,3 +34,18 @@ def json_encode(obj, indent=4):
 def load_file(path):
     with open(path, 'r') as fd:
         return json.load(fd)
+
+
+def json_loads(obj, keys=None):
```
Comment (Kami) [inline st2common/st2common/util/jsonify.py]: Would be nice to add docstrings which explains what function does :P

**Comment by jfryman** [comment]
Sorry, this might not be super clear... for the sake of discussion.

The ability to use dot notation from a JSON output is a goal, but equally there may be a scenario where I still need the raw JSON to pass to something else (say another webhook). Can I eat my :cake: and have it too?

File: st2common/st2common/util/jsonify.py (hunk 1 above)
Comment (lakshmi-kannan) [inline st2common/st2common/util/jsonify.py]: OMG.

**Comment by lakshmi-kannan** [comment]
jfryman: I think if shell actions actually output json string that means they want to push structured data to us and are taking the json.dumps route to send a string. So we converting them to json seems like a reasonable thing to do. Wdyt?

**Comment by jfryman** [comment]
@lakshmi-kannan yeah, thinking about that... I can appreciate that logic... especially if we're calling JSON our common currency. I think this works. :+1:

**Comment by enykeev** [comment]
An (unintended) consequence of that change is that now `core.local cmd="echo 1"` will return `1` as a number and not a string it's supposed to be.

A simple example of where it might lead us:
```
[
    {
        "action": "core.local",
        "callback": {},
        "context": {
            "user": "stanley"
        },
        "end_timestamp": "2015-01-26T05:19:50.575000Z",
        "id": "54c5ce760640fd6f3bbed1f9",
        "parameters": {
            "cmd": "echo 0.99999999999999999999999999"
        },
        "result": {
            "failed": false,
            "return_code": 0,
            "stderr": "",
            "stdout": 1.0,
            "succeeded": true
        },
        "start_timestamp": "2015-01-26T05:19:50.476000Z",
        "status": "succeeded"
    },
    {
        "action": "core.local",
        "callback": {},
        "context": {
            "user": "stanley"
        },
        "end_timestamp": "2015-01-26T05:19:37.552000Z",
        "id": "54c5ce690640fd6f3bbed1f8",
        "parameters": {
            "cmd": "echo 0.99999999999999"
        },
        "result": {
            "failed": false,
            "return_code": 0,
            "stderr": "",
            "stdout": 0.99999999999999,
            "succeeded": true
        },
        "start_timestamp": "2015-01-26T05:19:37.452000Z",
        "status": "succeeded"
    },
    ...
]
```

Remind me, what exactly was the reason for this change in the first place?

**Comment by Kami** [comment]
Yeah, good point.

I'm +1 for leaving stdout as it is and adding a result attribute (which could simply be parsed stdout). In fact, that's similar to what Python runner already does so it's good from the consistency perspective as well.

**Comment by manasdk** [comment]
> If we want to have a parsed version of the output, I suggest we introduce another field, "result" as James proposed, while keeping stdout a string (as it should be based on its name).

We already saved plenty of duplicated data not a big fan of adding a new property. If an action output structured data then that was the intent of the action and we should parse and preserve as structured data. This PR is basically in support of that view. Structured data is easy to work with in workflows etc. therefore we should *always* prefer structured data over plain old strings.

The fact that json.loads behavior as per the conversion table https://docs.python.org/2/library/json.html#json.JSONDecoder is an issue. There is some amount of surprise for the user in this behavior. We can limit conversions to objects and skip the other types by using a custom decoder. This can be an implementation compromise we could make.

**Comment by manasdk** [comment]
@Kami (Regarding python runner).

I guess you are referring to https://github.com/StackStorm/st2/blob/master/st2actions/st2actions/runners/pythonrunner.py#L167.

This means in an execution the result itself has a result property? How do we use these thing in workflows and rules where we tie the actions together?

**Comment by enykeev** [comment]
> We already saved plenty of duplicated data not a big fan of adding a new property.

I don't actually care whether we are going to save it or recalculate it on each request.

> If an action output structured data then that was the intent of the action and we should parse and preserve as structured data. This PR is basically in support of that view. Structured data is easy to work with in workflows etc. therefore we should always prefer structured data over plain old strings.

I don't see why you have decided that local runner would (and should) return structured output. The only thing we know with it for sure is that every call will result in a binary streams of stdout, stderr and a resulting return code. Trying to convert stdout automatically a) changes its semantical sense b) creates uncertainty of the structure (we don't know whether we are going to get a string, number or object and you have to check the type every time; you can't just str() in js) c) would require substantial effort to create custom decoder to avoid all the possible cases where it might do not what we are expecting it to do. But most importantly, we are getting string from the local runner and we are sending string to the user, why do we need all that chain of error-prone conversions in the process? 

I see it is useful for workflow. I get it and I think it's good thing we don't have to parse json for every workflow task, but I also see the number of problems we would have to deal with preparing, formating and showing all kinds of results to the user.

> The fact that json.loads behavior as per the conversion table https://docs.python.org/2/library/json.html#json.JSONDecoder is an issue. There is some amount of surprise for the user in this behavior.

I really don't see anything new here, this is how JSON decoder works in pretty much every language I have dealt with, `1`, `[]` and `null` were always been a valid JSON.

**Comment by manasdk** [comment]
> recalculate it on each request
This will most likely end up getting expensive.

**Comment by manasdk** [comment]
> I don't see why you have decided that local runner would (and should) return structured output.

This makes it easy for the output of a shell action, that is spitting out structured data, to be consumed in workflows, rules etc. Keep it as a string of trying to do a json loads in the workflow engine or rules engine is of course inefficient.

**Comment by manasdk** [comment]
> but I also see the number of problems we would have to deal with preparing, formating and showing all kinds of results to the user.

I honestly believe we want to solve these problems. The more such features we take in the platform the better it is for the user.

**Comment by manasdk** [comment]
> I really don't see anything new here, this is how JSON decoder works in pretty much every language I have dealt with, 1, [] and null were always been a valid JSON.

Yeah but I thought you said that "1" or "0.99" should be left alone as strings.

I do understand you were making a point about semantics and you do have a solid irrefutable point that by performing a json.loads we affect semantics of the result. When weighed against the benefit of being useful in a workflow I guess the changed semantics are ok since we are preserving the same data.

**Comment by Kami** [comment]
@manasdk 

>@Kami (Regarding python runner).

I>  guess you are referring to https://github.com/StackStorm/st2/blob/master/st2actions/st2actions/runners/pythonrunner.py#L167.

Yes

> This means in an execution the result itself has a result property? How do we use these thing in workflows and rules where we tie the actions together?

You should be able to access .result attribute.

> We already saved plenty of duplicated data not a big fan of adding a new property.

I personally don't think duplicated data is that big of a deal. If this ever becomes a problem we could also trade of some CPU for compression of stdout and stderr fields.

**Pull request merged**